### PR TITLE
Impl and test elbow method for automatic optimal k selection in k-means clustering

### LINKED ISSUE #20: Impl: clusterByDistance()
- #22

### PR DESCRIPTION
**Pull request by z-q-ying**
## Highlights
- Close #20
- Impl and test automatic clustering using the k-means algorithm; uses the elbow method to determine the optimal k (the best number of clusters)
- Functions implemented and tested in `kMeansClustering` and `groupingDistance.test`. The only required call is to `export function kMeansClusteringWrapper(ids: string[], items: any): string[][] {}` in other files, e.g., `grouping`

## Clustering rules:

- If there are fewer than 5 nodes (const, TBU), there will be only 1 cluster
- If, upon attempting to create a 2-cluster configuration, the distortion—as measured by the mean squared distance—is less than 200 (const, TBU), it is considered appropriate to propose a 2-cluster grouping
- Generally, k will range from 2 to maxK, where maxK = dataPoints.length / 3 + 1, assuming each cluster will generally have 3 or more items (again, 3 is a const, TBU)

### CODE REVIEW

File: test/mockBoardNodes.ts
Comment (z-q-ying) [inline test/mockBoardNodes.ts:1]: Auto-formatting only

**Comment by cptbtptp01** [comment]
for this algorithm, does it cover case like two items connect by a connector, they form a cluster regardless of distance?

**Comment by z-q-ying** [comment]
no, purely based on center coordinate (x, y); case of connectors will ideally be handled elsewhere but not in this cluster function

**Review by cptbtptp01 — APPROVED** [review]
Understood the logic after discussion, given related functions have been tested, this should be a good to go for plugging into the main grouping logic! thanks for the effort!

File: src/kMeansClustering.ts [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+    return DEFAULT_ONE_CLUSTER;
+  }
+
+  let maxK = dataPoints.length / ESTIMATED_NODES_PER_CLUSTER + 1;
+  let distortions = [];
+
+  for (let k = 2; k <= maxK; k++) {
+    let [centroids, assignments] = runKMeansForNTimes(dataPoints, k);
+    let distortion = getMeanSquaredDistance(dataPoints, centroids, assignments);
+    if (k === 2 && distortion < MAX_MSD_FOR_TWO_CLUSTER) {
```
[the remark this answers is not in the document]
Comment (z-q-ying) [reply src/kMeansClustering.ts:54]: not really, it handles the case where items can form only two clusters

**Pull request merged**